### ISSUE TITLE
[dev.icinga.com #8905] Reloading icinga schedules (unwanted) passive checks

### ISSUE THREAD

**Issue by icinga-migration** [Icinga/icinga2#2848]
This issue has been migrated from Redmine: https://dev.icinga.com/issues/8905

**Created by kruisdraad on 2015-03-30 12:29:24 +00:00**

Assignee: _(none)_
Status: _New_ 
Target Version: _(none)_
Last Update: _2016-11-09 14:59:55 +00:00 (in Redmine)_

    Icinga Version: r2.3.2-1
    Backport?: Not yet backported
    Include in Changelog: 1

---

At the moment of a icinga reload/restart it will attempt to schedule the initial check of all services. Thats fine for active checks, but passive checks report in on their own.

As the active check part is a fallback check (usually a dummy check with a WARN/CRIT value) and results in a hard state directly (3/3) you will get a LOT of alerts.

Ideally you want to 'define' those services as passive and use the normal scheduling *after* starting. e.g. if the check\_interval is 1m, then the first check would be 1m after startup.

enable\_passive\_checks is not documented and might be related in the handling of this 'event'

**Comment by Al2Klimov** [comment]
Hello @kruisdraad,

did you create the issue? (Unfortunately we can only guess.)

If yes, is this problem still present in v2.10.2?

Best,
AK

**Comment by kruisdraad** [comment]
@Al2Klimov yes i did make this, however i no longer operate the large monitoring setup where i can easily  reproduce it. I cannot help you at this time.

**Comment by Al2Klimov** [comment]
We re-schedule active and freshness checks the same way – which is a bit wrong IMAO:

https://github.com/Icinga/icinga2/blob/7ed690610fedde3e503b35f01ba68b3090496ca5/lib/icinga/checkable.cpp#L66-L70

I.e. on start if the next check would happen during the next minute, re-schedule it to a random point of time during the next minute. E.g. if the next check would happen at now+59s, re-schedule it to now+1s.

@N-o-X Shall we re-schedule them not earlier than they're scheduled at the moment?

**Comment by kyrias** [comment]
I can confirm that it is still happening with 2.11.3, which makes #4200 even more annoying as my workaround for it was to trigger a restart whenever we create new services through the API, but that leads to constant CRITICAL services until they report in again.

**Comment by kyrias** [comment]
Assuming that the reason for doing this rescheduling thing is to try to prevent too many checks from being clumped together to the same time, I think something like the following would be more reasonable:

```C++
double next_check = GetNextCheck();
double delta = std::min(GetCheckInterval(), 60.0); 
delta *= (double)std::rand() / RAND_MAX; 
if (next_check < now) {
	SetNextCheck(now + delta);
} else if (next_check < now + 60) {
	SetNextCheck(next_check + delta)
}
```

That way we never schedule a check earlier than it was already scheduled to be run, and we somewhat prevent them from being clumped up together on restarts.

@Al2Klimov @N-o-X Would something like that be acceptable?

**Comment by Al2Klimov** [comment]
_PING_ @N-o-X

**Comment by Al2Klimov** [comment]
> That way we never schedule a check earlier than it was already scheduled to be run

Hm... why not. But let's get a second opinion...

**Comment by kyrias** [comment]
> > That way we never schedule a check earlier than it was already scheduled to be run
> 
> Hm... why not. But let's get a second opinion...

Because it's 1) never useful to do so and 2) actively breaks the whole concept of passive checks.

But I don't have to maintain an Icinga instance anymore so I no longer care about this.

**Comment by julianbrost** [comment]
> That way we never schedule a check earlier than it was already scheduled to be run

Sounds like a good idea to me, doing so would only result in checking something more often than asked to. But I'd even go a step further: is there a reason to reschedule anything for which the next check is still in the future? I think it should be perfectly fine to keep any next check that's in the future as is and only reschedule checks that are due now so that they are spread out over some time instead of all running immediately.
